src/sagemaker/workflow/utilities.py

```diff
@@ -197,3 +197,32 @@ def wrapper(*args, **kwargs):
         return func(*args, **kwargs)
 
     return wrapper
+
+
+def valid_edges(adjacency_list: Dict[str, List[str]]) -> Set:
+    """Returns all valid edges to be used in a pipeline visualization
+
+    Deletes redundant edges through transitive reduction
+
+    Args:
+        adjacency_list (Dict[str, List[str]]): Adjacency list that represents the pipeline graph
+
+    Returns:
+        Set of valid edges
+
+    """
+    nodes = adjacency_list.keys()
+    edges = set()
+    for node in nodes:
+        out_bound_edges = adjacency_list[node]
+        for edge in out_bound_edges:
+            edges.add((node, edge))
+
+    for a in nodes:
+        for b in nodes:
```

Comment:
The nodes are a collection of string of names of steps right?
If so can we skip the loop early here?
like:
```
for node1 in nodes:
    for node2 in nodes:
        if node1 == node2:
            continue
        for node3 in nodes:
            if (node1 == node3) or (node2 == node3):
                continue
            if .....
                edges.remove(...)
```